feat(incremental): silent warning and add presets (#10175)
web-infra-dev/rspack: website/docs/en/config/experiments.mdx

<ApiMeta addedVersion="1.1.0-beta.0" />

- **Type:** `boolean | 'none' | 'safe' | 'advance' | 'advance-silent' | Incremental`
- **Default:** `'safe'`

> Whether to perform the rebuild incrementally to speed up the rebuild and HMR. Two configuration ways are provided:

Reviewer: ```suggestion
Whether to enable incremental build, which significantly speeds up rebuilds and HMR by only rebuilding the changed parts. Two configuration ways are provided:
```

> 1. Preset: including `boolean | 'none' | 'safe' | 'advance' | 'advance-silent'`

Reviewer: ```suggestion
1. Presets: including `boolean | 'none' | 'safe' | 'advance' | 'advance-silent'`
```

>    - `false | 'none'`: Turn off, and it will not be enabled for any stage.

Reviewer: ```suggestion
   - `false | 'none'`: Disable incremental build, and it will not be enabled for any stage.
```

   - `'safe'`: Enable incremental for `make` and `emitAssets` stages, This is also the current default behavior of Rspack.

>    - `true | 'advance-silent'`: Enable incremental for all stages to maximize the performance of rebuilds and HMR. After these stages become stable in the future, we will set this option as the default behavior of Rspack.

Reviewer: ```suggestion
   - `true | 'advance-silent'`: Enable incremental build for all stages to maximize performance for rebuilds and HMR. After these stages become stable in the future, we will make this option the default behavior for Rspack.
```

>    - `'advance'`: The same as the above, but it will throw a warning to prompt the user in cases where incremental is not friendly. This option serves as a supplement to `true | 'advance-silent'` to help users troubleshoot configurations that may potentially affect the performance of incremental rebuilds.

Reviewer: ```suggestion
   - `'advance'`: The same as above, but it will detect cases that are unfriendly to incremental build and throw warnings to users (e.g., incorrect configurations). This option can help you to identify potential issues affecting incremental build performance.
```

> 2. Detailed object configuration: `Incremental`, which enables fine-grained control over whether the incremental of each stage is enabled.

Reviewer: ```suggestion
2. **Detailed object configuration**: `Incremental`, which allows fine-grained control over whether the incremental is enabled for each stage.
```

   {/* prettier-ignore */}
   <details>
      <summary style={{ display: 'list-item' }}>`Incremental` 详细类型</summary>
      <blockquote>
        <p>
        ```ts
        type Incremental = {
          // Whether to throw a warning when encountering situations that are not friendly for incremental.
          silent?: boolean;
          // The following configuration is used to control whether the incremental of each stage is enabled.
          make?: boolean;
          inferAsyncModules?: boolean;
          providedExports?: boolean;
          dependenciesDiagnostics?: boolean;
          sideEffects?: boolean;
          buildChunkGraph?: boolean;
          moduleIds?: boolean;
          chunkIds?: boolean;
          modulesHashes?: boolean;
          modulesCodegen?: boolean;
          modulesRuntimeRequirements?: boolean;
          chunksRuntimeRequirements?: boolean;
          chunksHashes?: boolean;
          chunksRender?: boolean;
          emitAssets?: boolean;
        };
        ```
        </p>
      </blockquote>
    </details>

Usually, we recommend configuring in the preset way, and the detailed object configuration is only provided to facilitate bug troubleshooting.

> Incremental will only improve the speed of subsequent rebuilds and have no impact on the initial build. However, the initial build when the persistent cache is available will also be regarded as a rebuild, so as to use the incremental feature to speed up the build:

Reviewer: ```suggestion
Incremental only improves the rebuild performance and have no impact on the initial build. However, when persistent cache is available, initial builds are also treated as rebuilds too, and can benefit from incremental for performance.

The table below shows the results of incremental in different scenarios:
```
